### PR TITLE
PR: Verify if file is already shown in the combobox to avoid duplicates (Code Analysis)

### LINKED ISSUE #12883: File names are repeated in Code analysis file combobox
Instead of registering each file name once, we're doing it several times. This is how the problem looks graphically:

imagen

@juanis2112, please wait until @steff456 fix this one before filming the pending gif for our docs.

### PR DESCRIPTION
**Pull request by steff456**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Verify if the current file is already shown in the history combobox to avoid duplicates.

image

### Issue(s) Resolved

Fixes #12883 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: Steff456

### CODE REVIEW

**Comment by ccordoba12** [comment]
@CAM-Gerlach, could you help me to review @steff456's work? Thanks!

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Sure, in work now.

File: spyder/plugins/pylint/tests/test_pylint.py [hunk 1]
```diff
@@ -185,33 +202,37 @@ def test_pylint_widget_pylintrc(
                 for bad_name in bad_names])
 
 
-def test_pylint_max_history_conf(pylint_test_script, mocker, qtbot):
+def test_pylint_max_history_conf(
+    pylint_test_script, pylint_test_scripts, mocker, qtbot):
```
Comment (CAM-Gerlach) [inline spyder/plugins/pylint/tests/test_pylint.py:206]:
```suggestion
        pylint_test_script, pylint_test_scripts, mocker, qtbot):
```

PEP 8

File: spyder/plugins/pylint/widgets/pylintgui.py [hunk 2]
```diff
@@ -258,12 +259,24 @@ def set_filename(self, filename):
...
+                self.filecombo.addItem(filename)
+                self.filecombo.setCurrentIndex(self.filecombo.count()-1)
+            else:
+                self.filecombo.setCurrentIndex(
+                    self.filecombo.findText(filename))
+            self.curr_filenames.append(filename)
+
+            is_parent = self.parent is not None
+            if is_parent and (self.filecombo.count() >
+                    self.parent.get_option('max_entries')):
```
Comment (CAM-Gerlach) [inline spyder/plugins/pylint/widgets/pylintgui.py:273]:
```suggestion
            if is_parent and (self.filecombo.count() >
                              self.parent.get_option('max_entries')):

```

PEP 8

File: spyder/plugins/pylint/plugin.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ class Pylint(SpyderPluginWidget):
     def __init__(self, parent=None):
         SpyderPluginWidget.__init__(self, parent)
 
-        max_entries = self.get_option('max_entries', 50)
+        max_entries = self.get_option('max_entries', 20)
```
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/pylint/plugin.py:52]: I agree. Let's use a constant to define that.

File: spyder/plugins/pylint/plugin.py [hunk 4]
```diff
@@ -132,7 +132,7 @@ def change_history_depth(self):
         depth, valid = QInputDialog.getInt(self, _('History'),
                                        _('Maximum entries'),
                                        self.get_option('max_entries'),
-                                       10, 10000)
+                                       5, 20)
```
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/pylint/plugin.py:135]: I agree with also using a constant for min and max. Although I don't like such a big value for the max (999). Why not just 100?

[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/pylint/plugin.py:135]: > Although I don't like such a big value for the max (999).

Could you explain why not, sorry? Is there a technical reason?

> Why not just 100?

Right back at ya, why not just 999 :)

A user might work with a large number of files, or want to go way back in their history, or not mind seeing a larger number of files, or we might want to increase it for testing, or.... We go to the effort of offering the setting, so unless there is a technical reason to restrict it, I don't see why we should set an arbitrary limit that's not well above any remotely reasonable bound and well below anything likely to scale poorly (> 1000s - 10 000s). The user can always change it again if they don't like seeing that many list entries.

**Comment by CAM-Gerlach** [comment]
Also, I realize this wasn't the original UX, but if its straightforward to do as part of this change, the ordering of the dialog should be reversed so that the newest files are displayed at the top, to match the standard convention for these comboboxes that users expect, and how the other similar ones (e.g. Find in Files) work.

File: spyder/plugins/pylint/tests/test_pylint.py [hunk 5]
```diff
@@ -86,12 +86,7 @@ def pylint_test_script(pylintrc_search_paths):
     with open(script_path, mode="w",
               encoding="utf-8", newline="\n") as script_file:
         script_file.write(PYLINT_TEST_SCRIPT)
-    return script_path
 
-
-@pytest.fixture
```
[the remark this answers is not in the document]
Comment (steff456) [reply spyder/plugins/pylint/tests/test_pylint.py]: I need to have more than one file, to verify that they are added as expected in the combobox

[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/pylint/tests/test_pylint.py]: Sure thing, that's what the code above does :) You can either use a fixture factory along the lines of what I've suggested above, which takes a list of filenames and returns multiple scripts (as I demonstrate in the example call, which returns the paths of three scripts just as the current code does), or just one that returns a single script, which you can call as many times as you need (once for each script).

File: spyder/plugins/pylint/widgets/pylintgui.py [hunk 6]
```diff
@@ -258,12 +260,36 @@ def set_filename(self, filename):
...
-        self.filecombo.selected()
+            self.filecombo.setCurrentIndex(
+                self.filecombo.findText(filename))
+
+    def change_history_limit(self, new_limit):
+        """Change the number of files listed in the history combobox."""
+        if self.filecombo.count() > new_limit:
+            num_elements = self.filecombo.count()
+            diff = num_elements - new_limit
+            for i in range(diff):
```
Comment (CAM-Gerlach) [inline spyder/plugins/pylint/widgets/pylintgui.py:288]:
```suggestion
            for __ in range(diff):
```

More clearly indicate this variable is unused.

File: spyder/plugins/pylint/tests/test_pylint.py [hunk 7]
```diff
@@ -185,33 +202,47 @@ def test_pylint_widget_pylintrc(
                 for bad_name in bad_names])
 
 
-def test_pylint_max_history_conf(pylint_test_script, mocker, qtbot):
+def test_pylint_max_history_conf(pylint_test_scripts, mocker, qtbot):
```
Comment (CAM-Gerlach) [inline spyder/plugins/pylint/tests/test_pylint.py:205]:
```suggestion
def test_pylint_max_history_conf(pylint_test_scripts, mocker):
```

Qtbot is no longer used

**Comment by CAM-Gerlach** [comment]
For some reason, my original review when missing...here's the remaining relevant portion.

* [ ] I was able to reproduce the following error with this PR, which occurred with any file already on the list on Spyder startup, and which did not occur on the same files when tested in Spyder 4.1.3:

1. Select any Python file to analyze using the file browser option that's already in the list of analyzed files
2. No file path appears in the combobox
3. Click analyze and the following error traceback is printed in a modal popup dialog, which has no context or user friendly help text:

image

<details>

```
  File "C:\Miniconda3\envs\spyder-test\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Miniconda3\envs\spyder-test\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\__main__.py", line 18, in <module>
    pylint.run_pylint()
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\__init__.py", line 22, in run_pylint
    PylintRun(sys.argv[1:])
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\lint\run.py", line 338, in __init__
    linter.check(args)
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\lint\pylinter.py", line 871, in check
    self.get_ast, self._iterate_file_descrs(files_or_modules)
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\lint\pylinter.py", line 903, in _check_files
    for name, filepath, modname in file_descrs:
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\lint\pylinter.py", line 960, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules):
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\lint\pylinter.py", line 977, in _expand_files
    self.add_message(key, args=message)
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\message\message_handler_mix_in.py", line 237, in add_message
    message_definition, line, node, args, confidence, col_offset
  File "C:\Miniconda3\envs\spyder-test\lib\site-packages\pylint\message\message_handler_mix_in.py", line 290, in add_one_message
    self.stats["by_module"][self.current_name][msg_cat] += 1
KeyError: None
```

</details>

I'm presuming something goes wrong with the check that the file is not already in the list, resulting in no filename actually being passed to the analysis.

**Comment by goanpeca** [comment]
This is probably a controversial opinion, but I will air it anyway since this has been going back and forth for a while.

I honestly do not see the real need for having an option to set how many items should or should not appear on a dropdown. As a user I do not really need to care about this, and if I do now I need to think about what number makes sense for a minimum or a maximum.

This is not going to solve the issue for Spyder 4.x since it is a UI breaking change, but I see no reason at all to have this option for Spyder 5.

**Comment by CAM-Gerlach** [comment]
Thanks @steff456 for the amazing work! @ccordoba12 ready for a final review :)

**Comment by ccordoba12** [comment]
> In the spirit of simplifying the UI I honestly do not see the real need for having an option to set how many items should or should not appear on a dropdown. As a user I do not really need to care about this, and if I do now I need to think about what number makes sense for a minimum or a maximum.

I understand it's an extra bit of configurability, but 

* Heavy users of this plugin (like @CAM-Gerlach) find it very useful, so I don't see anything terribly bad in leaving it.
* It's quite clear what it does now, whereas before it was quite confusing.

So I'm -1 on removing it, even for Spyder 5.

**Pull request merged**